ext/CEGUI/IrrlichtBaWTexture.cpp

```diff
@@ -0,0 +1,167 @@
+        return { bufferRowLength ? bufferRowLength : params.extent.width, params.extent.height };
+    }
+    else 
+    {
+        assert(status);
+        return { 0, 0 };
+    }
+}
+
+const CEGUI::Vector2f& IrrlichtBaWTexture::getTexelScaling() const
+{
+    bool status = gpuTexture.get();
+
+    if (status)
+    {
+        auto& params = gpuTexture->getCreationParameters().image->getCreationParameters();
+        auto width = bufferRowLength ? bufferRowLength : params.extent.width;
+
+        return
+        {
+            static_cast<float>(width) == 0.f ? 0.f : 1.f / static_cast<float>(width),
+            static_cast<float>(params.extent.height) == 0.f ? 0.f : 1.f / static_cast<float>(params.extent.height)
+        };
+    }
+    else
+    {
+        assert(status);
+        return { 0, 0 };
+    }
+}
+
+void IrrlichtBaWTexture::loadFromFile(const CEGUI::String& filename, const CEGUI::String& resourceGroup)
+{
+    auto path = resourceGroup + filename;
+    auto assetManager = device->getAssetManager();
+
+    auto params = asset::IAssetLoader::SAssetLoadParams();
+    auto bundle = assetManager->getAsset(path, params);
+    auto cpuImage = asset::IAsset::castDown<asset::ICPUImage>(bundle.getContents().first[0]);
+    assert(cpuImage.get(), "something went wrong while casting, empty pointer detected!");
+
+    bufferRowLength = cpuImage->getRegions().begin()->bufferRowLength;
+
+    asset::ICPUImageView::SCreationParams imageViewInfo;
+    imageViewInfo.image = std::move(cpuImage);
+    imageViewInfo.format = imageViewInfo.image->getCreationParameters().format;
+    imageViewInfo.viewType = asset::IImageView<asset::ICPUImage>::ET_2D;
+    imageViewInfo.flags = static_cast<asset::ICPUImageView::E_CREATE_FLAGS>(0u);
+    imageViewInfo.subresourceRange.baseArrayLayer = 0u;
+    imageViewInfo.subresourceRange.baseMipLevel = 0u;
+    imageViewInfo.subresourceRange.layerCount = imageViewInfo.image->getCreationParameters().arrayLayers;
+    imageViewInfo.subresourceRange.levelCount = imageViewInfo.image->getCreationParameters().mipLevels;
+
+    auto cpuImageView = asset::ICPUImageView::create(std::move(imageViewInfo));
+
+    // cachingName = ""     TODO Okay, so we have to get likely GL id or something similar. Should I cast it to OpenGL image or maybe we should get CPU cache string?
+
+    gpuTexture = device->getVideoDriver()->getGPUObjectsFromAssets(&cpuImageView.get(), &cpuImageView.get() + 1u)->front();
+    
+}
+
+void IrrlichtBaWTexture::loadFromMemory(const void* buffer, const CEGUI::Sizef& buffer_size, PixelFormat pixel_format)
+{
+    asset::E_FORMAT newFormat = [&]()
+    {
+        switch (pixel_format)
+        {
+            case PixelFormat::PF_RGB: return asset::EF_R8G8B8_SRGB;
+            case PixelFormat::PF_RGBA: return asset::EF_R8G8B8A8_SRGB;
+            case PixelFormat::PF_RGBA_4444: return asset::EF_R4G4B4A4_UNORM_PACK16;
+            case PixelFormat::PF_RGB_565: return asset::EF_R5G6B5_UNORM_PACK16;
```

Comment:
don't support these, they are not SRGB, and CEGUI probably gets them wrong

Reply:
adjusted in following commit